src/io/avro/read/deserialize.rs

```diff
@@ -123,7 +123,16 @@ fn deserialize_value<'a>(
                 .downcast_mut::<DynMutableListArray<i32>>()
                 .unwrap();
             loop {
-                let len = util::zigzag_i64(&mut block)? as usize;
+                let len = util::zigzag_i64(&mut block)?;
+                let len = if len < 0 {
+                    // Avro spec: If a block's count is negative, its absolute value is used,
+                    // and the count is followed immediately by a long block size indicating the number of bytes in the block. This block size permits fast skipping through data, e.g., when projecting a record to a subset of its fields.
+                    let _ = util::zigzag_i64(&mut block)?;
+
+                    -len
+                } else {
+                    len
+                };
```

Comment:
This PR looks correct to me, but it actually fixes a different bug than the one I was referring to in #1252. The fix here addresses the handling the case when an avro list item block encodes an optional `byte_size` for data skipping as `<-1*block_count><byte_size>...`.

The issue I was referring to, however, is the call to  `array.try_push_valid()?;` on line 145 (136 original), which is called for *each* block deserialised. Shouldn't this call be moved right outside of the `loop {...}` onto line 147, so that we only push valid *once* per list item? 

Thinking out loud, for an empty list, looks like the `if len == 0 { break; }` causes this loop to short circuit currently meaning we won't push a valid. I wonder if this is also an unintended bug, of having this call to try_push_valid in the block unrolling loop as opposed to outside.

To clarify, I am concerned about the handling of the following two cases:

- `List [1,2,3,4]`, encoded as two "blocks"
  - Currently we make **2 calls** to `try_push_valid` whereas I would **expect 1**

- `List []`, encoded as 0 blocks, indicated by a prefixed byte of `len == 0`
  - Currently we make **0 calls** to `try_push_valid` whereas I would **expect 1**, as you mentioned before that empty lists/structs are still "valid" in Arrow so as to maintain O(1) validity checks.

Again, I am not well versed in the Avro format, so feel free to correct me if I have an incorrect understanding here. I came across these potential bugs while reading the deserialisation code thoroughly, as I was considering taking a stab at implementing Map type support for Avro, which will be very similar to `List`. Thank you

Reply:
Agree. Really good call.

Did another push with an extra fix

Reply:
Awesome, LGTM